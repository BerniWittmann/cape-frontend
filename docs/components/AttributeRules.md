# AttributeRules

The component to allow to view and edit the Attributes and Rules for the States of a Context Factor

## Props

<!-- @vuese:AttributeRules:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|contextFactor|gets the Context Factor|`Object`|`true`|-|

<!-- @vuese:AttributeRules:props:end -->


## Methods

<!-- @vuese:AttributeRules:methods:start -->
|Method|Description|Parameters|
|---|---|---|
|convertToTableData|Converts the context rules to a representation that can be used in the table|-|
|convertFromTableData|Converts the context rules table representation to a string that can be stored|-|
|reRender|Triggers a new render of the component|-|

<!-- @vuese:AttributeRules:methods:end -->


