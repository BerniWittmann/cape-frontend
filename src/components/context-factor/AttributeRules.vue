<template>
  <div class="attr-rules-margin">
    <div v-if="attrExists">
      <el-row :gutter="20">
        <el-form ref='newStateForm' :model="newState" label-position="top" :rules="rules" status-icon inline
                 @submit.native.prevent>
          <el-form-item prop="stateName">
            <el-input class="input" ref="stateInput" size="small"
                      v-model="newState.stateName"
                      :placeholder="$t('context_factor.edit.states.name')"
            ></el-input>
          </el-form-item>
          <el-form-item>
            <el-button class="b-margin" icon="el-icon-plus" @click="addNewState" size="small">
              {{ $t('context_factor.edit.states.add') }}
            </el-button>
          </el-form-item>
        </el-form>
      </el-row>
      <el-row :gutter="20">
        <el-table :data="tableData" border>

          <el-table-column
                  v-for="(ca, index) in contextFactor.attributes" :prop="ca.id"
                  align="center" :key="index" :resizable="false"
                  :min-width="200">
            <template slot="header" slot-scope="scope">
              <span v-if="show">{{ ca.key }}</span>
            </template>
            <template slot-scope="scope">
              <RulesCell v-if="show" :value="scope.row[ca.id]" :type="ca.type"
                         @change="(nv) => {scope.row[ca.id] = nv}"></RulesCell>
            </template>
          </el-table-column>

          <el-table-column
                  :label="$t('context_factor.edit.state')"
                  sortable
                  :sort-method=nameCompare
                  fixed="right"
                  align="center"
                  :min-width="150"
                  :resizable="false"
                  class-name="state-column">
            <template slot-scope="scope">
              <div class="parent">
                <input-edit class="state-input" :value="scope.row.state" @change="(nst) => {scope.row.state = nst}"
                            :rules="rules.stateName"
                            size="mini"></input-edit>
                <span class="btn">
              <el-button size="mini" icon="el-icon-delete" type="danger" class="delete-btn"
                         @click="deleteState(scope.$index)"></el-button>
            </span></div>
            </template>
          </el-table-column>
        </el-table>
      </el-row>
    </div>
    <div v-else>
      <i class="el-icon-warning"></i>
      <span>{{ $t('context_factor.edit.states.no_attributes') }}</span>
    </div>
  </div>
</template>

<script>
/* ============
 * AttributeRules
 * ============
 *
 *  The component to allow to view and edit the Attributes and Rules for the States of a Context Factor
 */

import RulesCell from '@/components/context-factor/RulesCell'
import InputEdit from '@/components/InputEdit.vue'

/*
 * @vuese
 * @group Components/Context-Factor
 *
 * The component to allow to view and edit the Attributes and Rules for the States of a Context Factor
 */

export default {
  name: 'AttributeRules',
  components: {
    RulesCell,
    InputEdit
  },

  props: {
    // gets the Context Factor
    contextFactor: {
      type: Object,
      required: true
    }
  },

  data() {
    return {
      // Data in Backend
      newState: {
        stateName: undefined
      },
      rules: {
        stateName: [
          { required: true, message: this.$t('context_factor.edit.validation.stateName.required'), trigger: 'blur' }
        ]
      },
      // Data for Table used in Frontend, after Parsing
      tableData: [],
      show: true
    }
  },

  computed: {},

  methods: {
    attrExists() {
      return this.contextFactor.attributes.length > 0
    },

    addNewState() {
      this.$refs.newStateForm.validate((valid) => {
        if (valid) {
          let newRow = { state: this.newState.stateName }
          for (let ca of this.contextFactor.attributes) {
            newRow[ca.id] = ''
          }
          this.tableData.push(newRow)
          this.newState.stateName = ''
        }
      })
    },

    convertToTableData() {
      // on tab click
      if (this.contextFactor.contextRules !== undefined) {
        let tableD = []
        for (let i = 0; i < this.contextFactor.contextRules.length; i++) {
          let row = { state: this.contextFactor.contextRules[i].state }
          for (let cr of this.contextFactor.contextRules[i].rule.split('&&')) {
            // possible separators are: == or <= or >=
            if (cr.indexOf('==') >= 0) {
              let ct = cr.split('==')[0]
              row[ct] = cr.split('==')[1]
            } else {
              if (cr.indexOf('<=') >= 0) {
                let ct = cr.split('<=')[0]
                row[ct] = (row[ct] !== undefined) ? row[ct] + cr.split('<=')[1] : '-' + cr.split('<=')[1]
              }
              if (cr.indexOf('>=') >= 0) {
                let ct = cr.split('>=')[0]
                row[ct] = (row[ct] !== undefined) ? cr.split('>=')[1] + row[ct] : cr.split('>=')[1] + '-'
              }
            }
          }
          tableD.push(row)
        }
        this.tableData = tableD
      }
    },

    nameCompare(a, b) {
      if (!a || !a.state || !b || !b.state) return 0
      const lca = a.state.toLowerCase()
      const lcb = b.state.toLowerCase()
      return lca > lcb ? 1 : (lca < lcb ? -1 : 0)
    },

    convertFromTableData() {
      // on save and tab click
      let cRules = []
      for (let cs of this.tableData) {
        let cr = {
          state: cs.state,
          rule: ''
        }
        for (let ca of this.contextFactor.attributes) {
          if (cs[ca.id] !== undefined && cs[ca.id].length > 0) {
            if (cr.rule.length > 0) {
              cr.rule += '&&'
            }
            switch (ca.type) {
              case 'Boolean':
                cr.rule += ca.id + '==' + cs[ca.id]
                break
              case 'String':
                cr.rule += ca.id + '==' + cs[ca.id]
                break
              default:
                let sPos = cs[ca.id].indexOf('-')
                if (sPos < 0) {
                  if (cs[ca.id] + '' !== 'undefined') {
                    cr.rule += ca.id + '==' + cs[ca.id]
                  }
                } else {
                  let ct = cs[ca.id].split('-')
                  if (ct[0] === ct[1]) {
                    if (ct[0] + '' !== 'undefined') {
                      cr.rule += ca.id + '==' + ct[0]
                    }
                  } else {
                    let k = 0
                    for (let i = 0; i < ct.length; i++) {
                      if (ct[i].length > 0) {
                        if (k > 0) {
                          cr.rule += '&&'
                        }
                        if (cs[ca.id].indexOf(ct[i]) < sPos) {
                          // >=
                          if (ct[i] + '' !== 'undefined') {
                            cr.rule += ca.id + '>=' + ct[i]
                          }
                          k++
                        } else {
                          // <=
                          if (ct[i] + '' !== 'undefined') {
                            cr.rule += ca.id + '<=' + ct[i]
                          }
                          k++
                        }
                      }
                    }
                  }
                }
            }
          }
        }
        cRules.push(cr)
      }
      this.contextFactor.contextRules = cRules
    },

    deleteState(stateIndex) {
      this.$confirm(this.$t('context_factor.edit.states.delete.message'), this.$t('context_factor.edit.states.delete.warning'), {
        confirmButtonText: this.$t('context_factor.edit.states.delete.ok'),
        cancelButtonText: this.$t('context_factor.edit.states.delete.cancel'),
        type: 'warning',
        cancelButtonClass: 'is-plain el-button--info',
        confirmButtonClass: 'el-button--danger'
      }).then(() => {
        this.tableData.splice(stateIndex, 1)
        this.convertFromTableData()
        this.$message({
          type: 'success',
          message: this.$t('context_factor.edit.states.delete.confirmation')
        })
      }).catch(() => {
        this.$message({
          type: 'info',
          message: this.$t('context_factor.edit.states.delete.cancellation')
        })
      })
    },

    reRender() {
      // needed to ensure that data shown is updated
      this.show = false
      this.$nextTick(() => {
        this.show = true
        this.$nextTick(() => {
        })
      })
    }
  },

  beforeMount() {
    this.convertToTableData()
  }
}
</script>

<style scoped lang="scss">

.attr-rules-margin {
  margin-left: 10px;
  margin-right: 10px;
}

.state-input {
  width: 100%;
}

.parent {
  display: flex;
}

.btn {
  display: inline-block;
  text-align: right;
  align-self: center;
}

.delete-btn {
  padding: 4px;
}

</style>
<style>
.el-table .state-column {
  background-color: #f5f6fa;
}
</style>
