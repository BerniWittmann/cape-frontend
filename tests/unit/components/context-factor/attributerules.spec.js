import { shallowMount } from '@vue/test-utils'
import { i18n } from '../../setupPlugins'

import AttributeRules from '@/components/context-factor/AttributeRules.vue'

describe('Components', () => {
  describe('AttributeRules', () => {
    let cmp
    let props
    beforeEach(() => {
      props = {
        contextFactor: {
          id: 'cF1',
          name: 'My COntext Factor',
          contextRules: [{
            id: 'cr1',
            state: 'First State',
            rule: 'ca1==OK&&ca2=19&&ca3==TRUE'
          }, {
            id: 'cr2',
            state: 'Second State',
            rule: 'ca3==FALSE'
          }],
          attributes: [{
            id: 'ca1',
            key: 'Attr 1',
            type: 'String',
            value: 'OK'
          }, {
            id: 'ca2',
            key: 'Attr 2',
            type: 'Number',
            value: '12'
          }, {
            id: 'ca3',
            key: 'Attr 3',
            type: 'Boolean',
            value: 'TRUE'
          }]
        }
      }
      jest.clearAllMocks()
      render()
    })

    function render() {
      cmp = shallowMount(AttributeRules, {
        i18n,
        propsData: props
      })
    }

    it('renders', () => {
      expect(cmp.html()).toMatchSnapshot()
    })

    it('renders a table', () => {
      const table = cmp.find('eltable-stub')
      expect(table.exists()).toBeTruthy()
      const columns = cmp.findAll('eltablecolumn-stub')
      expect(columns.length).toEqual(4)
      expect(columns.at(0).props('prop')).toEqual('ca1')
      expect(columns.at(1).props('prop')).toEqual('ca2')
      expect(columns.at(2).props('prop')).toEqual('ca3')
      expect(columns.at(3).props('label')).toEqual('context_factor.edit.state')
    })

    it('does not render a table if no attributes exist', () => {
      props.contextFactor = { ...props.contextFactor, attributes: [] }
      cmp.setProps(props)
      expect(cmp.html()).toMatchSnapshot()
    })

    it('can add a new rule', () => {
      cmp.vm.$refs.newStateForm.validate = jest.fn().mockImplementation((arg) => arg(true))
      const input = cmp.find('elinput-stub')
      expect(input.exists()).toBeTruthy()
      cmp.vm.newState.stateName = 'New State'
      const btn = cmp.find('elbutton-stub')
      expect(btn.exists()).toBeTruthy()
      btn.vm.$emit('click')

      expect(cmp.vm.$refs.newStateForm.validate).toHaveBeenCalled()
      expect(cmp.html()).toMatchSnapshot()
      expect(cmp.vm.tableData.length).toEqual(3)
      expect(cmp.vm.tableData[2]).toEqual({ ca1: undefined, ca2: undefined, ca3: undefined, state: 'New State' })
    })

    it('can add a new rule', () => {
      cmp.vm.newState.stateName = 'New &|=.State'
      cmp.vm.checkStateName()
      expect(cmp.vm.newState.stateName).toEqual('New State')
    })

    it('does not add new rule on invalid input', () => {
      cmp.vm.$refs.newStateForm.validate = jest.fn().mockImplementation((arg) => arg(false))
      const btn = cmp.find('elbutton-stub')
      expect(btn.exists()).toBeTruthy()
      btn.vm.$emit('click')

      expect(cmp.html()).toMatchSnapshot()
      expect(cmp.vm.tableData.length).toEqual(2)
    })

    describe('can delete a state', () => {
      it('can delete a state after confirmation', () => {
        cmp.vm.$message = jest.fn()
        cmp.vm.convertFromTableData = jest.fn()
        cmp.vm.$confirm = jest.fn().mockReturnValue(({ then: (arg) => { arg(); return { catch: () => {} } } }))
        cmp.vm.deleteState(0)
        expect(cmp.vm.$confirm).toHaveBeenCalledWith('context_factor.edit.states.delete.message', 'context_factor.edit.states.delete.warning', expect.any(Object))
        expect(cmp.vm.convertFromTableData).toHaveBeenCalled()
        expect(cmp.vm.tableData.length).toEqual(1)
        expect(cmp.vm.$message).toHaveBeenCalledWith({ message: 'context_factor.edit.states.delete.confirmation', type: 'success' })
      })
      it('can abort the state deletion', () => {
        cmp.vm.$confirm = jest.fn().mockReturnValue(({ then: () => { return { catch: (arg) => { arg() } } } }))
        cmp.vm.$message = jest.fn()
        cmp.vm.convertFromTableData = jest.fn()
        cmp.vm.deleteState(0)
        expect(cmp.vm.$confirm).toHaveBeenCalledWith('context_factor.edit.states.delete.message', 'context_factor.edit.states.delete.warning', expect.any(Object))
        expect(cmp.vm.convertFromTableData).not.toHaveBeenCalled()
        expect(cmp.vm.tableData.length).toEqual(2)
        expect(cmp.vm.$message).toHaveBeenCalledWith({ message: 'context_factor.edit.states.delete.cancellation', type: 'info' })
      })
    })

    describe('it exposes methods', () => {
      it('reRender', () => {
        const before = cmp.html()
        cmp.vm.reRender()
        expect(cmp.html()).toEqual(before)
      })

      it('nameCompare', () => {
        expect(cmp.vm.nameCompare(undefined, undefined)).toEqual(0)
        expect(cmp.vm.nameCompare({}, {})).toEqual(0)
        expect(cmp.vm.nameCompare({ state: 'a' }, {})).toEqual(0)
        expect(cmp.vm.nameCompare({}, { state: 'b' })).toEqual(0)
        expect(cmp.vm.nameCompare({}, { state: 'b' })).toEqual(0)
        expect(cmp.vm.nameCompare({ state: 'A' }, { state: 'b' })).toEqual(-1)
        expect(cmp.vm.nameCompare({ state: 'z' }, { state: 'T' })).toEqual(1)
      })

      describe('convertFromTableData', () => {
        function convert(rules) {
          cmp.vm.tableData = rules
          cmp.vm.convertFromTableData()
        }
        it('can handle empty Table Data', () => {
          convert([])
          expect(cmp.vm.contextFactor.contextRules).toEqual([])
        })

        describe('can handle Strings', () => {
          it('parses Strings', () => {
            convert([{ state: 'My State', ca1: 'OK' }])
            expect(cmp.vm.contextFactor.contextRules).toEqual([{ state: 'My State', rule: 'ca1==OK' }])
          })
        })

        describe('can handle Booleans', () => {
          it('parses Booleans', () => {
            convert([{ state: 'My State', ca3: 'TRUE' }])
            expect(cmp.vm.contextFactor.contextRules).toEqual([{ state: 'My State', rule: 'ca3==TRUE' }])
          })
        })

        describe('can handle Numbers', () => {
          it('handles equality', () => {
            convert([{ state: 'My State', ca2: '12' }])
            expect(cmp.vm.contextFactor.contextRules).toEqual([{ state: 'My State', rule: 'ca2==12' }])

            convert([{ state: 'My State', ca2: 'undefined' }])
            expect(cmp.vm.contextFactor.contextRules).toEqual([{ state: 'My State', rule: '' }])

            convert([{ state: 'My State', ca2: '12_12' }])
            expect(cmp.vm.contextFactor.contextRules).toEqual([{ state: 'My State', rule: 'ca2==12' }])
          })

          it('handles greater or equal comparison', () => {
            convert([{ state: 'My State', ca2: '12_' }])
            expect(cmp.vm.contextFactor.contextRules).toEqual([{ state: 'My State', rule: 'ca2>=12' }])
          })
          it('handles lesser or equal comparison', () => {
            convert([{ state: 'My State', ca2: '_12' }])
            expect(cmp.vm.contextFactor.contextRules).toEqual([{ state: 'My State', rule: 'ca2<=12' }])
          })
          describe('handles intervals', () => {
            it('handles correct intervals', () => {
              convert([{ state: 'My State', ca2: '12_14' }])
              expect(cmp.vm.contextFactor.contextRules).toEqual([{ state: 'My State', rule: 'ca2>=12&&ca2<=14' }])
            })
            it('handles undefined values', () => {
              convert([{ state: 'My State', ca2: 'undefined_undefined' }])
              expect(cmp.vm.contextFactor.contextRules).toEqual([{ state: 'My State', rule: '' }])

              convert([{ state: 'My State', ca2: 'undefined_12' }])
              expect(cmp.vm.contextFactor.contextRules).toEqual([{ state: 'My State', rule: '&&ca2<=12' }])

              convert([{ state: 'My State', ca2: '41_undefined' }])
              expect(cmp.vm.contextFactor.contextRules).toEqual([{ state: 'My State', rule: 'ca2>=41&&' }])
            })
          })
        })

        it('parses connections', () => {
          convert([{ state: 'My State', ca1: 'OK', ca3: 'TRUE' }])
          expect(cmp.vm.contextFactor.contextRules).toEqual([{ state: 'My State', rule: 'ca1==OK&&ca3==TRUE' }])
        })
      })

      describe('convertToTableData', () => {
        function setRules(rules) {
          props.contextFactor = { ...props.contextFactor, contextRules: rules }
          cmp.setProps(props)
        }
        it('if no rules are present return nothing', () => {
          const before = cmp.vm.tableData
          setRules(undefined)
          cmp.vm.convertToTableData()
          expect(cmp.vm.tableData).toEqual(before)
        })

        it('if empty rules are present return nothing', () => {
          setRules([])
          cmp.vm.convertToTableData()
          expect(cmp.vm.tableData).toEqual([])
        })

        it('handle equality', () => {
          setRules([{ state: 'First', rule: 'a1==12' }])
          cmp.vm.convertToTableData()
          expect(cmp.vm.tableData).toEqual([{ state: 'First', a1: '12' }])

          setRules([{ state: 'First', rule: 's5==STRING' }])
          cmp.vm.convertToTableData()
          expect(cmp.vm.tableData).toEqual([{ state: 'First', s5: 'STRING' }])
        })

        it('handle greater or equal comparison', () => {
          setRules([{ state: 'First', rule: 'a1>=12' }])
          cmp.vm.convertToTableData()
          expect(cmp.vm.tableData).toEqual([{ state: 'First', a1: '12_' }])
        })

        it('handle lesser or equal comparison', () => {
          setRules([{ state: 'First', rule: 'a1<=12' }])
          cmp.vm.convertToTableData()
          expect(cmp.vm.tableData).toEqual([{ state: 'First', a1: '_12' }])

          setRules([{ state: 'First', rule: 'a1==10&&a1<=12' }])
          cmp.vm.convertToTableData()
          expect(cmp.vm.tableData).toEqual([{ state: 'First', a1: '1012' }])
        })

        it('handle interval', () => {
          setRules([{ state: 'First', rule: 'a1<=10&&a1>=12' }])
          cmp.vm.convertToTableData()
          expect(cmp.vm.tableData).toEqual([{ state: 'First', a1: '12_10' }])

          setRules([{ state: 'First', rule: 'a1>=10&&a1<=12' }])
          cmp.vm.convertToTableData()
          expect(cmp.vm.tableData).toEqual([{ state: 'First', a1: '10_12' }])
        })

        it('handle connection', () => {
          setRules([{ state: 'First', rule: 'a1<=10&&a1>=12&&b1==TRUE&&c1==19' }])
          cmp.vm.convertToTableData()
          expect(cmp.vm.tableData).toEqual([{ state: 'First', a1: '12_10', b1: 'TRUE', c1: '19' }])
        })

        it('handle multiple rules', () => {
          setRules([{ state: 'First', rule: 'a1<=10&&a1>=12' }, { state: 'Second', rule: 'c1==TRUE&&b1>=19' }])
          cmp.vm.convertToTableData()
          expect(cmp.vm.tableData).toEqual([{ state: 'First', a1: '12_10' }, { state: 'Second', c1: 'TRUE', b1: '19_' }])
        })
      })
    })
  })
})
