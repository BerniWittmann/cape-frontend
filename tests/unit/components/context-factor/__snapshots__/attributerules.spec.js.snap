// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components AttributeRules can add a new rule 1`] = `
<div class="attr-rules-margin">
  <div>
    <elrow-stub tag="div" gutter="20" justify="start" align="top">
      <elform-stub model="[object Object]" rules="[object Object]" labelposition="top" labelsuffix="" inline="true" statusicon="true" showmessage="true" validateonrulechange="true">
        <elformitem-stub prop="stateName" inlinemessage="" showmessage="true">
          <elinput-stub size="small" type="text" validateevent="true" placeholder="context_factor.edit.states.name" class="input" value=""></elinput-stub>
        </elformitem-stub>
        <elformitem-stub inlinemessage="" showmessage="true">
          <elbutton-stub type="default" size="small" icon="el-icon-plus" nativetype="button" class="b-margin">
            context_factor.edit.states.add
          </elbutton-stub>
        </elformitem-stub>
      </elform-stub>
    </elrow-stub>
    <elrow-stub tag="div" gutter="20" justify="start" align="top">
      <eltable-stub data="[object Object],[object Object]" fit="true" border="true" showheader="true" selectonindeterminate="true">
        <eltablecolumn-stub type="default" prop="ca1" minwidth="200" align="center" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
        <eltablecolumn-stub type="default" prop="ca2" minwidth="200" align="center" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
        <eltablecolumn-stub type="default" prop="ca3" minwidth="200" align="center" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
        <eltablecolumn-stub type="default" label="context_factor.edit.state" classname="state-column" minwidth="150" sortable="" sortmethod="function () { [native code] }" align="center" fixed="right" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
      </eltable-stub>
    </elrow-stub>
  </div>
</div>
`;

exports[`Components AttributeRules does not add new rule on invalid input 1`] = `
<div class="attr-rules-margin">
  <div>
    <elrow-stub tag="div" gutter="20" justify="start" align="top">
      <elform-stub model="[object Object]" rules="[object Object]" labelposition="top" labelsuffix="" inline="true" statusicon="true" showmessage="true" validateonrulechange="true">
        <elformitem-stub prop="stateName" inlinemessage="" showmessage="true">
          <elinput-stub size="small" type="text" autocomplete="off" validateevent="true" placeholder="context_factor.edit.states.name" class="input"></elinput-stub>
        </elformitem-stub>
        <elformitem-stub inlinemessage="" showmessage="true">
          <elbutton-stub type="default" size="small" icon="el-icon-plus" nativetype="button" class="b-margin">
            context_factor.edit.states.add
          </elbutton-stub>
        </elformitem-stub>
      </elform-stub>
    </elrow-stub>
    <elrow-stub tag="div" gutter="20" justify="start" align="top">
      <eltable-stub data="[object Object],[object Object]" fit="true" border="true" showheader="true" selectonindeterminate="true">
        <eltablecolumn-stub type="default" prop="ca1" minwidth="200" align="center" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
        <eltablecolumn-stub type="default" prop="ca2" minwidth="200" align="center" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
        <eltablecolumn-stub type="default" prop="ca3" minwidth="200" align="center" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
        <eltablecolumn-stub type="default" label="context_factor.edit.state" classname="state-column" minwidth="150" sortable="" sortmethod="function () { [native code] }" align="center" fixed="right" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
      </eltable-stub>
    </elrow-stub>
  </div>
</div>
`;

exports[`Components AttributeRules does not render a table if no attributes exist 1`] = `
<div class="attr-rules-margin">
  <div><i class="el-icon-warning"></i> <span>context_factor.edit.states.no_attributes</span></div>
</div>
`;

exports[`Components AttributeRules renders 1`] = `
<div class="attr-rules-margin">
  <div>
    <elrow-stub tag="div" gutter="20" justify="start" align="top">
      <elform-stub model="[object Object]" rules="[object Object]" labelposition="top" labelsuffix="" inline="true" statusicon="true" showmessage="true" validateonrulechange="true">
        <elformitem-stub prop="stateName" inlinemessage="" showmessage="true">
          <elinput-stub size="small" type="text" autocomplete="off" validateevent="true" placeholder="context_factor.edit.states.name" class="input"></elinput-stub>
        </elformitem-stub>
        <elformitem-stub inlinemessage="" showmessage="true">
          <elbutton-stub type="default" size="small" icon="el-icon-plus" nativetype="button" class="b-margin">
            context_factor.edit.states.add
          </elbutton-stub>
        </elformitem-stub>
      </elform-stub>
    </elrow-stub>
    <elrow-stub tag="div" gutter="20" justify="start" align="top">
      <eltable-stub data="[object Object],[object Object]" fit="true" border="true" showheader="true" selectonindeterminate="true">
        <eltablecolumn-stub type="default" prop="ca1" minwidth="200" align="center" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
        <eltablecolumn-stub type="default" prop="ca2" minwidth="200" align="center" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
        <eltablecolumn-stub type="default" prop="ca3" minwidth="200" align="center" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
        <eltablecolumn-stub type="default" label="context_factor.edit.state" classname="state-column" minwidth="150" sortable="" sortmethod="function () { [native code] }" align="center" fixed="right" filtermultiple="true" sortorders="ascending,descending,"></eltablecolumn-stub>
      </eltable-stub>
    </elrow-stub>
  </div>
</div>
`;
